### PR TITLE
[SWE-24] Forgot Password Backend

### PR DESCRIPTION
**Pull request by sammyoto**
INCOMPLETE. Added the basic code for implementing a password reset flow with amplify and cognito. The issue is our users are not required to input an email in order to sign up. We need to redo the create account and make users sign up with email. I know Derek asked me to do that earlier this week, just didn't get to it yet. Apologies.
Screen Shot 2023-10-16 at 4 30 47 PM
Also I keep getting this error for some reason.

### CODE REVIEW

File: pages/forgot-password.tsx [hunk 1]
```diff
@@ -2,6 +2,7 @@ import Head from 'next/head';
 import isEmail from 'validator/lib/isEmail';
 import { useRouter } from 'next/router';
 import React, { useState } from 'react';
+import { Auth } from 'aws-amplify';
 import styles from '../styles/Home.module.css';
 
 function ForgotPassword () {
```
Comment (jackfales) [inline pages/forgot-password.tsx:8]:
```suggestion
function ForgotPassword() {
```

File: pages/reset-password.tsx [hunk 2]
```diff
@@ -0,0 +1,79 @@
+import Head from 'next/head';
+import isEmail from 'validator/lib/isEmail';
+import { useRouter } from 'next/router';
+import React, { useState } from 'react';
+import { Auth } from 'aws-amplify';
+import styles from '../styles/Home.module.css';
+
+function ResetPassword () {
```
Comment (jackfales) [inline pages/reset-password.tsx:8]:
```suggestion
function ResetPassword() {
```

File: pages/forgot-password.tsx [hunk 3]
```diff
@@ -25,9 +26,15 @@ function ForgotPassword () {
         setErrorMessage("")
     }
 
-    console.log(email)
+    // Send a request to initiate the password reset process
+    // Send confirmation code to user's email
```
Comment (jackfales) [inline pages/forgot-password.tsx:30]: Is this comment still supposed to be there? Where is the logic to send the confirmation code to a user's email?

File: pages/reset-password.tsx [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  const [errorMessage, setErrorMessage] = useState('Please enter your email, verification code sent to your email, and your new password');
+
+  const router = useRouter();
+
+  const onSubmit = async (event) => {
+    event.preventDefault();
+
+    // Get form data
```
Comment (jackfales) [inline pages/reset-password.tsx:16]: No need for this comment you can tell what the code is doing already

File: pages/reset-password.tsx [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        </div>
+      </form>
+    </div>
+  );
+};
+
+export default ResetPassword;
+  
+
+
```
Comment (jackfales) [inline pages/reset-password.tsx:79]:
```suggestion
```

**Review by jackfales — COMMENTED** [review]
LGTM couple of changes but overall looks good, just do some testing

**Comment by sammyoto** [comment]
If everything works correctly you should now be able to:
- create an account
- verify the account with a confirmation code sent to email
- reset password with forgot password

File: pages/create-account.js [hunk 6]
```diff
@@ -10,17 +12,19 @@ export default function CreateAccount() {
     lastName: '',
     password: '',
     username: '',
+    email: ''
```
Comment (datderek) [inline pages/create-account.js:15]: Nit - Jack got me for the same thing last week, but lets move the email variable to after confirmPassword. Just to maintain alphabetic consistency.

File: pages/reset-password.tsx [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+          <h3>Username</h3>
+          <input type="text" name="username" />
+        </div>
+        <div id="verificationCode">
+          <h3>Verification Code</h3>
+          <input type="text" name="verificationCode" />
+        </div>
+        <div id="newPassword">
+          <h3>New Password</h3>
+          <input type="text" name="newPassword" />
```
Comment (datderek) [inline pages/reset-password.tsx:55]:
```suggestion
          <input type="password" name="newPassword" />
```

File: pages/reset-password.tsx [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        </div>
```
Comment (datderek) [inline pages/reset-password.tsx:56]: May want to also add a field to confirm new password with some error handling - like what we have in `create-account.js`.

**Review by datderek — COMMENTED** [review]
Tested, was able to create account, reset password, and login with new password. Had some small comments.

File: pages/create-account.js
Comment (jackfales) [inline pages/create-account.js:1]: nit: consider converting this file to TypeScript. Not required this PR but if it's simple might as well

File: pages/create-account.js [hunk 9]
```diff
@@ -33,13 +37,18 @@ export default function CreateAccount() {
...
+            'Please provide a valid email!' : '';
+          break;
         case 'username':
           errorMessage = input.length < 5 ? 
             'Username must be at least 5 characters long!' : '';
           break;
         case 'password':
-          errorMessage = input.length < 5 ?
-            'Password must be at least 6 characters long!' : '';
+          const passwordRegex = /^(?=.*[0-9!@#$%^&*])[A-Za-z0-9!@#$%^&*]{6,}$/;
```
Comment (jackfales) [inline pages/create-account.js:49]: This works, but is there a library that can test the password for us? If not, add a comment describing what the regex is checking for

File: pages/create-account.js [hunk 10]
```diff
@@ -61,29 +70,36 @@ export default function CreateAccount() {
     // Check if the form has any errors
     const hasNoErrors = Object.values(errors)
                                 .every((input) => input === '');
+
+    let username = formData.get("username")
+    let password = formData.get("password")
+    let email = formData.get("email")
+    let firstName = formData.get("firstName")
+    let lastName = formData.get("lastName")
```
Comment (jackfales) [inline pages/create-account.js:78]: Change to `const`

File: pages/create-account.js [hunk 11]
```diff
@@ -61,29 +70,36 @@ export default function CreateAccount() {
...
-            setErrorMessages({ serverResponse: res.message });
+      try {
+        await Auth.signUp({
+          username,
+          password,
+          attributes: {
+            email: email,
+            given_name: firstName,         
+            family_name: lastName,
+            preferred_username: username
```
Comment (jackfales) [inline pages/create-account.js:89]: nit: Arrange attributes with some form of ordering (alphabetical or by purpose)

File: pages/reset-password.tsx [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+  const onSubmit = async (event) => {
+    event.preventDefault();
+
+    const formData = new FormData(event.target)
+    
+    const username = formData.get("username").toString()
+    const verificationCode = formData.get("verificationCode").toString()
+    const newPassword = formData.get("newPassword").toString()
+
+    const passwordRegex = /^(?=.*[0-9!@#$%^&*])[A-Za-z0-9!@#$%^&*]{6,}$/;
```
Comment (jackfales) [inline pages/reset-password.tsx:21]: Same comment on regex as above

File: pages/reset-password.tsx [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+    // Make sure password conforms to regex
+    if (!passwordRegex.test(newPassword)) {
+      setErrorMessage("Password must be at least 6 characters long and contain one special character!")
+    } else{
```
Comment (jackfales) [inline pages/reset-password.tsx:26]:
```suggestion
    } else {
```

File: pages/verify-email.tsx [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+    const [errorMessage, setErrorMessage] = useState('Please input your username and confirmation code sent to your email.');
+    const router = useRouter();
+
+    const handleSubmit = async (e) => {
+        e.preventDefault();
+
+        const formData = new FormData(e.target);
+
+        let username = formData.get("username").toString()
+        let confirmationCode = formData.get("confirmationCode").toString()
```
Comment (jackfales) [inline pages/verify-email.tsx:17]: Change to const

**Review by jackfales — APPROVED** [review]
Couple of nitpick comments but overall LGTM, good work

**Comment by sammyoto** [comment]
Changed the password policy in cognito to minimum of 8 characters, 1 number, 1 special character. Used validator library to validate password.

File: pages/create-account.tsx [hunk 15]
```diff
@@ -1,26 +1,39 @@
...
   });
 
+  const passwordConstraints = {
+    minLength: 8,        // Minimum length of 8 characters
+    minLowercase: 0,      // Minimum 0 lowercase letters
+    minUppercase: 0,      // Minimum 0 uppercase letters
+    minNumbers: 1,       // Minimum 1 numeric digits
+    minSymbols: 1,       // Minimum 1 special character
+  };
+
```
Comment (datderek) [inline pages/create-account.tsx:26]: Remove comments, variable names are already self-explanatory.

File: pages/reset-password.tsx [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+
+function ResetPassword() {
+  const [errorMessage, setErrorMessage] = useState('Please enter your email, verification code sent to your email, and your new password');
+
+  const passwordConstraints = {
+    minLength: 8,        // Minimum length of 8 characters
+    minLowercase: 0,      // Minimum 0 lowercase letters
+    minUppercase: 0,      // Minimum 0 uppercase letters
+    minNumbers: 1,       // Minimum 1 numeric digits
+    minSymbols: 1,       // Minimum 1 special character
```
Comment (datderek) [inline pages/reset-password.tsx:16]: Delete comments, variable names self-explanatory

File: pages/create-account.tsx [hunk 17]
```diff
@@ -1,26 +1,39 @@
...
   const router = useRouter();
 
   const handleSubmit = async (e) => {
     e.preventDefault();
 
-    const userInputs = {};
-    const errors = {};
     const formData = new FormData(e.target);
 
+    const userInputs = {};
```
Comment (datderek) [inline pages/create-account.tsx:34]: Can remove as it looks like userInputs is no longer being used

Comment (datderek) [reply pages/create-account.tsx:34]: Should also be another line of userInputs underneath the `switch` statement as well.

File: pages/create-account.tsx [hunk 18]
```diff
@@ -56,38 +73,50 @@ export default function CreateAccount() {
...
 
     // Check if the form has any errors
     const hasNoErrors = Object.values(errors)
                                 .every((input) => input === '');
+
+    const username = formData.get("username").toString()
+    const password = formData.get("password").toString()
+    const email = formData.get("email")
+    const firstName = formData.get("firstName")
+    const lastName = formData.get("lastName")
```
Comment (datderek) [inline pages/create-account.tsx:92]: Any reason only `username` and `password` need the `toString()` method?

File: pages/reset-password.tsx [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+  const router = useRouter();
+
+  const onSubmit = async (event) => {
+    event.preventDefault();
+
+    const formData = new FormData(event.target)
+    
+    const username = formData.get("username").toString()
+    const verificationCode = formData.get("verificationCode").toString()
+    const newPassword = formData.get("newPassword").toString()
```
Comment (datderek) [inline pages/reset-password.tsx:28]: Are `toString()` needed?

**Review by datderek — APPROVED** [review]
Couple of nits and clean up, then it should be good to go!

File: pages/create-account.tsx (hunk 18 above)
Comment (sammyoto) [reply pages/create-account.tsx:92]: formData.get() returns a type of formDataEntryValue. Typescript now enforces the type for the Auth methods to string. Doesn't seem to care about the attributes, but does care about username and password. I'll just make them all strings for consistency.

**Review by datderek — APPROVED** [review]
Looks good to me.

**Pull request merged**